src/extension.c

```diff
@@ -156,16 +170,23 @@ extension_set_state(enum ExtensionState newstate)
 static void
 extension_update_state()
 {
-	static bool in_recursion = false;
-	/* Since the state of the extension is determined by the snapshot of the transaction there
-	 * is no point processing recursive calls as the outer call will always set the correct state.
-	 * This also prevents deep recursion during `AcceptInvalidationMessages`.
+	enum ExtensionState new_state = extension_current_state();
+
+	/* Never actually set the state to "not installed" since there is no good
+	 * way to get out of it in case the extension is installed again in
+	 * another backend. After the extension has been dropped, the proxy table
+	 * no longer exists and when the extension is reinstalled, the proxy table
+	 * will have a different relid. Therefore, there is no way to identify the
+	 * invalidation on the proxy table when CREATE EXTENSION is issued in
+	 * another backend. Nor is it allowed to lookup the new relid in the
+	 * invalidation callback, since that may lead to bad behavior.
+	 *
+	 * Instead, set the state to "unknown" so that a "slow path" lookup of the
+	 * actual state has to be made next time the state is queried.
 	 */
-	if (in_recursion)
-		return;
+	if (new_state == EXTENSION_STATE_NOT_INSTALLED)
+		new_state = EXTENSION_STATE_UNKNOWN;
```

Comment:
I considered removing the "not installed" state, but decided against it for now to minimize changes (also, this is a loader change). `extension_current_state()` actually still determines that the extension is not installed and can therefore distinguish between the cases. It could be useful for debugging and the only reason this state isn't set right now is because I couldn't figure out a good way to signal other backends to get out of it when the extension is created again.